src/lib/geoip/city-approximation.ts

```diff
@@ -0,0 +1,139 @@
+import fs from 'node:fs';
+import got from 'got';
+import AdmZip from 'adm-zip';
+import csvParser from 'csv-parser';
+import _ from 'lodash';
+import { getRedisClient, RedisClient } from '../redis/client.js';
+import throttle from '../ws/helper/throttle.js';
+import { scopedLogger } from '../logger.js';
+import { getIsDcCity } from './dc-cities.js';
+
+type City = {
+	geonameId: string
+	name: string
+	asciiName: string
+	alternateNames: string
+	latitude: string
+	longitude: string
+	featureClass: string
+	featureCode: string
+	countryCode: string
+	cc2: string
+	admin1Code: string
+	admin2Code: string
+	admin3Code: string
+	admin4Code: string
+	population: number
+	elevation: string
+	dem: string
+	timezone: string
+	modificationDate: string
+}
+
+type CsvCityRow = City & { population: string };
+
+const logger = scopedLogger('city-approximation');
+
+export const URL = 'https://download.geonames.org/export/dump/cities15000.zip';
+const FILENAME = 'GEONAMES_CITIES.csv';
+
+const query = async (url: string): Promise<Buffer> => {
+	const result = await got(url, {
+		responseType: 'buffer',
+		timeout: { request: 5000 },
+	});
+
+	return result.body;
+};
+
+export const updateGeonamesCitiesFile = async (): Promise<void> => {
+	const response = await query(URL);
+	const zip = new AdmZip(response);
+	zip.extractEntryTo('cities15000.txt', 'data', false, true, false, FILENAME);
+};
+
+let redis: RedisClient;
+let geonamesCities: Map<string, City> = new Map();
+
+export const getCity = async (city = '', country?: string, latitude?: number, longitude?: number) => {
+	if (city === '') { return city; }
+
+	const isDcCity = getIsDcCity(city, country);
+
+	if (isDcCity) { return city; }
+
+	const approximatedCity = await getApproximatedCity(country, latitude, longitude);
+
+	if (!approximatedCity) { return city; }
+
+	return approximatedCity;
+};
+
+export const populateCitiesList = async () => {
+	redis = getRedisClient();
+	const cities = await readCitiesCsvFile() as CsvCityRow[];
+
+	geonamesCities = new Map(cities.map(city => ([ city.geonameId, { ...city, population: parseInt(city.population, 10) }])));
+
+	await Promise.all(cities.map(async (city: City) => {
```

Comment:
It's better to use `Bluebird.map()` (http://bluebirdjs.com/docs/api/promise.map.html) with limited `concurrency` for looping over large lists, otherwise, 25k promises get spawn at the same time here, which may cause various random issues with the whole process.